Read local and S3 oracle output target data

annakrystalli opened the pull request:
This PR builds on #77 & #78 and resolves #72 and #76 by introducing functionality to create an appropriate schema and read oracle output target data from a local and s3 cloud hub.

I also noticed a setting that needed to be included when opening csv datasets (`quoted_na = TRUE`) to ensure `NA`s are read in correctly as `NA`s and not `""`  which I noticed was happening in some instances. SO I applied this to `connect_target_timeseries()` too.

Note that so far functionality is basic, focusing only on reading, and I will next move on to developing more robust validation functions for each target data type.

annakrystalli commented:
In the end I couldn't really figure out how to get round the failing windows test when using a local SubTreeFileSystem to mimic an s3  with the following note/justification:

>   Skipping the test on windows because the use of lower level
     `SubTreeFileSystem$create()` function on windows is throwing unrelated errors
    related to local paths on windows which are not guaranteed to work:
    (see <https://arrow.apache.org/docs/cpp/api/filesystem.html#subtree-filesystem-wrapper>)
    We've already shown these tests work on linux and macos and on actual s3 cloud hubs
    with single files.

If we wanted to be really robust with our testing, we could add an additional s3 example hub with multi-file target data structure.

zkamvar reviewed (commented):
This looks great, thank you! 

I agree with the choice to skip the windows tests.

WRT to the `quoted_na` argument, this will mean that any hubs that use a task ID value of "NA" (e.g. the two-letter code for north america) will have problems because suddenly those values will be converted to missing and may not be scored properly.

In our documentation, we [state that missing data in a CSV need to be a blank cell or NA with no quotes](https://hubverse.io/en/latest/user-guide/model-output.html#example-model-output-as-csv), so it's codified what is expected. I think the default of `quoted_na = FALSE` is the more sensible solution because if they have a missing value that's quoted, that's either a mistake (that should be corrected) or it's an actual meaningful value.

annakrystalli commented:
> In our documentation, we [state that missing data in a CSV need to be a blank cell or NA with no quotes](https://hubverse.io/en/latest/user-guide/model-output.html#example-model-output-as-csv), so it's codified what is expected. I think the default of `quoted_na = FALSE` is the more sensible solution because if they have a missing value that's quoted, that's either a mistake (that should be corrected) or it's an actual meaningful value.

Hmmm Houston I think we have a problem. If we don't set `quoted_na = TRUE`, blank cells are not converted to `NA`s. See below:
``` r
tmp_hub_path <- withr::local_tempdir()
example_hub <- "https://github.com/hubverse-org/example-complex-forecast-hub.git"
gert::git_clone(url = example_hub, path = tmp_hub_path)


oracl <-  arrow::read_csv_arrow(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"))
# Add quoted NA value to location
oracl$location[1] <- "NA"
# Turn NA to empty string in output_type_id
oracl$output_type_id[is.na(oracl$output_type_id)] <- ""
readr::write_csv(oracl, fs::path(tmp_hub_path, "target-data", "oracle-output.csv"))

# With quoted NA = FALSE, empty strings are not recognised as NAs
arrow::open_dataset(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"),
                    format = "csv",
                    schema = hubData::create_oracle_output_schema(tmp_hub_path),
                    skip = 1) |>
  dplyr::collect()
#> # A tibble: 200,340 × 6
#>    location target_end_date target       output_type output_type_id oracle_value
#>    <chr>    <date>          <chr>        <chr>       <chr>                 <dbl>
#>  1 NA       2022-10-22      wk inc flu … quantile    ""                     2380
#>  2 01       2022-10-22      wk inc flu … quantile    ""                      141
#>  3 02       2022-10-22      wk inc flu … quantile    ""                        3
#>  4 04       2022-10-22      wk inc flu … quantile    ""                       22
#>  5 05       2022-10-22      wk inc flu … quantile    ""                       50
#>  6 06       2022-10-22      wk inc flu … quantile    ""                      124
#>  7 08       2022-10-22      wk inc flu … quantile    ""                       15
#>  8 09       2022-10-22      wk inc flu … quantile    ""                        9
#>  9 10       2022-10-22      wk inc flu … quantile    ""                        1
#> 10 11       2022-10-22      wk inc flu … quantile    ""                        8
#> # ℹ 200,330 more rows

# With quoted NA = TRUE, empty strings recognised as NAs but so is "NA"
arrow::open_dataset(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"),
                    format = "csv",
                    schema = hubData::create_oracle_output_schema(tmp_hub_path),
                    skip = 1, quoted_na = TRUE) |>
  dplyr::collect()
#> # A tibble: 200,340 × 6
#>    location target_end_date target       output_type output_type_id oracle_value
#>    <chr>    <date>          <chr>        <chr>       <chr>                 <dbl>
#>  1 <NA>     2022-10-22      wk inc flu … quantile    <NA>                   2380
#>  2 01       2022-10-22      wk inc flu … quantile    <NA>                    141
#>  3 02       2022-10-22      wk inc flu … quantile    <NA>                      3
#>  4 04       2022-10-22      wk inc flu … quantile    <NA>                     22
#>  5 05       2022-10-22      wk inc flu … quantile    <NA>                     50
#>  6 06       2022-10-22      wk inc flu … quantile    <NA>                    124
#>  7 08       2022-10-22      wk inc flu … quantile    <NA>                     15
#>  8 09       2022-10-22      wk inc flu … quantile    <NA>                      9
#>  9 10       2022-10-22      wk inc flu … quantile    <NA>                      1
#> 10 11       2022-10-22      wk inc flu … quantile    <NA>                      8
#> # ℹ 200,330 more rows

readr::read_csv(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"))
#> Rows: 200340 Columns: 6
#> ── Column specification ────────────────────────────────────────────────────────
#> Delimiter: ","
#> chr  (4): location, target, output_type, output_type_id
#> dbl  (1): oracle_value
#> date (1): target_end_date
#> 
#> ℹ Use `spec()` to retrieve the full column specification for this data.
#> ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
#> # A tibble: 200,340 × 6
#>    location target_end_date target       output_type output_type_id oracle_value
#>    <chr>    <date>          <chr>        <chr>       <chr>                 <dbl>
#>  1 <NA>     2022-10-22      wk inc flu … quantile    <NA>                   2380
#>  2 01       2022-10-22      wk inc flu … quantile    <NA>                    141
#>  3 02       2022-10-22      wk inc flu … quantile    <NA>                      3
#>  4 04       2022-10-22      wk inc flu … quantile    <NA>                     22
#>  5 05       2022-10-22      wk inc flu … quantile    <NA>                     50
#>  6 06       2022-10-22      wk inc flu … quantile    <NA>                    124
#>  7 08       2022-10-22      wk inc flu … quantile    <NA>                     15
#>  8 09       2022-10-22      wk inc flu … quantile    <NA>                      9
#>  9 10       2022-10-22      wk inc flu … quantile    <NA>                      1
#> 10 11       2022-10-22      wk inc flu … quantile    <NA>                      8
#> # ℹ 200,330 more rows
```

<sup>Created on 2025-04-02 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

Not sure what to suggest here. Note that the `quoted_na = TRUE` behaviour is the default for `readr::read_csv` too.

annakrystalli commented:
I did some more digging and it looks like there is no real way to combine `"NA"` and `NA` successfully.  The only way is to use `""` as `NA`s and specify that as the NA value when reading.

``` r
tmp_hub_path <- withr::local_tempdir()
example_hub <- "https://github.com/hubverse-org/example-complex-forecast-hub.git"
gert::git_clone(url = example_hub, path = tmp_hub_path)
# By default when using open_dataset with csv format, quoted_na = FALSE does not recognise NAs.
arrow::open_dataset(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"),
                    format = "csv",
                    schema = hubData::create_oracle_output_schema(tmp_hub_path),
                    skip = 1) |>
  dplyr::collect()
#> # A tibble: 200,340 × 6
#>    location target_end_date target       output_type output_type_id oracle_value
#>    <chr>    <date>          <chr>        <chr>       <chr>                 <dbl>
#>  1 US       2022-10-22      wk inc flu … quantile    NA                     2380
#>  2 01       2022-10-22      wk inc flu … quantile    NA                      141
#>  3 02       2022-10-22      wk inc flu … quantile    NA                        3
#>  4 04       2022-10-22      wk inc flu … quantile    NA                       22
#>  5 05       2022-10-22      wk inc flu … quantile    NA                       50
#>  6 06       2022-10-22      wk inc flu … quantile    NA                      124
#>  7 08       2022-10-22      wk inc flu … quantile    NA                       15
#>  8 09       2022-10-22      wk inc flu … quantile    NA                        9
#>  9 10       2022-10-22      wk inc flu … quantile    NA                        1
#> 10 11       2022-10-22      wk inc flu … quantile    NA                        8
#> # ℹ 200,330 more rows

arrow::open_dataset(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"),
                    format = "csv",
                    schema = hubData::create_oracle_output_schema(tmp_hub_path),
                    skip = 1, quoted_na = TRUE) |>
  dplyr::collect()
#> # A tibble: 200,340 × 6
#>    location target_end_date target       output_type output_type_id oracle_value
#>    <chr>    <date>          <chr>        <chr>       <chr>                 <dbl>
#>  1 US       2022-10-22      wk inc flu … quantile    <NA>                   2380
#>  2 01       2022-10-22      wk inc flu … quantile    <NA>                    141
#>  3 02       2022-10-22      wk inc flu … quantile    <NA>                      3
#>  4 04       2022-10-22      wk inc flu … quantile    <NA>                     22
#>  5 05       2022-10-22      wk inc flu … quantile    <NA>                     50
#>  6 06       2022-10-22      wk inc flu … quantile    <NA>                    124
#>  7 08       2022-10-22      wk inc flu … quantile    <NA>                     15
#>  8 09       2022-10-22      wk inc flu … quantile    <NA>                      9
#>  9 10       2022-10-22      wk inc flu … quantile    <NA>                      1
#> 10 11       2022-10-22      wk inc flu … quantile    <NA>                      8
#> # ℹ 200,330 more rows



# Add quoted NA value to location ----
oracl <-  arrow::read_csv_arrow(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"))
oracl$location[1] <- "NA"
readr::write_csv(oracl, fs::path(tmp_hub_path, "target-data", "oracle-output.csv"))

# With an "NA" value in location, quoted NA = FALSE does not recognise it as NA
# but nor does it other NAs
arrow::open_dataset(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"),
                    format = "csv",
                    schema = hubData::create_oracle_output_schema(tmp_hub_path),
                    skip = 1, quoted_na = FALSE) |>
  dplyr::collect()
#> # A tibble: 200,340 × 6
#>    location target_end_date target       output_type output_type_id oracle_value
#>    <chr>    <date>          <chr>        <chr>       <chr>                 <dbl>
#>  1 NA       2022-10-22      wk inc flu … quantile    NA                     2380
#>  2 01       2022-10-22      wk inc flu … quantile    NA                      141
#>  3 02       2022-10-22      wk inc flu … quantile    NA                        3
#>  4 04       2022-10-22      wk inc flu … quantile    NA                       22
#>  5 05       2022-10-22      wk inc flu … quantile    NA                       50
#>  6 06       2022-10-22      wk inc flu … quantile    NA                      124
#>  7 08       2022-10-22      wk inc flu … quantile    NA                       15
#>  8 09       2022-10-22      wk inc flu … quantile    NA                        9
#>  9 10       2022-10-22      wk inc flu … quantile    NA                        1
#> 10 11       2022-10-22      wk inc flu … quantile    NA                        8
#> # ℹ 200,330 more rows
# With an "NA" value in location, quoted NA = TRUE does recognise valid NAs but
# also converts "NA" to NA
arrow::open_dataset(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"),
                    format = "csv",
                    schema = hubData::create_oracle_output_schema(tmp_hub_path),
                    skip = 1, quoted_na = TRUE) |>
  dplyr::collect()
#> # A tibble: 200,340 × 6
#>    location target_end_date target       output_type output_type_id oracle_value
#>    <chr>    <date>          <chr>        <chr>       <chr>                 <dbl>
#>  1 <NA>     2022-10-22      wk inc flu … quantile    <NA>                   2380
#>  2 01       2022-10-22      wk inc flu … quantile    <NA>                    141
#>  3 02       2022-10-22      wk inc flu … quantile    <NA>                      3
#>  4 04       2022-10-22      wk inc flu … quantile    <NA>                     22
#>  5 05       2022-10-22      wk inc flu … quantile    <NA>                     50
#>  6 06       2022-10-22      wk inc flu … quantile    <NA>                    124
#>  7 08       2022-10-22      wk inc flu … quantile    <NA>                     15
#>  8 09       2022-10-22      wk inc flu … quantile    <NA>                      9
#>  9 10       2022-10-22      wk inc flu … quantile    <NA>                      1
#> 10 11       2022-10-22      wk inc flu … quantile    <NA>                      8
#> # ℹ 200,330 more rows
# Turn single NA value to empty string in output_type_id ----
oracl$output_type_id[1] <- ""
readr::write_csv(oracl, fs::path(tmp_hub_path, "target-data", "oracle-output.csv"))

# With quoted NA = FALSE, empty strings are not recognised as NAs
arrow::open_dataset(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"),
                    format = "csv",
                    schema = hubData::create_oracle_output_schema(tmp_hub_path),
                    skip = 1) |>
  dplyr::collect()
#> # A tibble: 200,340 × 6
#>    location target_end_date target       output_type output_type_id oracle_value
#>    <chr>    <date>          <chr>        <chr>       <chr>                 <dbl>
#>  1 NA       2022-10-22      wk inc flu … quantile    ""                     2380
#>  2 01       2022-10-22      wk inc flu … quantile    "NA"                    141
#>  3 02       2022-10-22      wk inc flu … quantile    "NA"                      3
#>  4 04       2022-10-22      wk inc flu … quantile    "NA"                     22
#>  5 05       2022-10-22      wk inc flu … quantile    "NA"                     50
#>  6 06       2022-10-22      wk inc flu … quantile    "NA"                    124
#>  7 08       2022-10-22      wk inc flu … quantile    "NA"                     15
#>  8 09       2022-10-22      wk inc flu … quantile    "NA"                      9
#>  9 10       2022-10-22      wk inc flu … quantile    "NA"                      1
#> 10 11       2022-10-22      wk inc flu … quantile    "NA"                      8
#> # ℹ 200,330 more rows

# With quoted NA = TRUE, empty strings recognised as NAs but so is "NA"
arrow::open_dataset(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"),
                    format = "csv",
                    schema = hubData::create_oracle_output_schema(tmp_hub_path),
                    skip = 1, quoted_na = TRUE, na = c("", "NA")) |>
  dplyr::collect()
#> # A tibble: 200,340 × 6
#>    location target_end_date target       output_type output_type_id oracle_value
#>    <chr>    <date>          <chr>        <chr>       <chr>                 <dbl>
#>  1 <NA>     2022-10-22      wk inc flu … quantile    <NA>                   2380
#>  2 01       2022-10-22      wk inc flu … quantile    <NA>                    141
#>  3 02       2022-10-22      wk inc flu … quantile    <NA>                      3
#>  4 04       2022-10-22      wk inc flu … quantile    <NA>                     22
#>  5 05       2022-10-22      wk inc flu … quantile    <NA>                     50
#>  6 06       2022-10-22      wk inc flu … quantile    <NA>                    124
#>  7 08       2022-10-22      wk inc flu … quantile    <NA>                     15
#>  8 09       2022-10-22      wk inc flu … quantile    <NA>                      9
#>  9 10       2022-10-22      wk inc flu … quantile    <NA>                      1
#> 10 11       2022-10-22      wk inc flu … quantile    <NA>                      8
#> # ℹ 200,330 more rows

# With quoted NA = TRUE, empty strings recognised as NAs but so is "NA"
arrow::open_dataset(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"),
                    format = "csv",
                    schema = hubData::create_oracle_output_schema(tmp_hub_path),
                    skip = 1, quoted_na = TRUE, na = c("")) |>
  dplyr::collect()
#> # A tibble: 200,340 × 6
#>    location target_end_date target       output_type output_type_id oracle_value
#>    <chr>    <date>          <chr>        <chr>       <chr>                 <dbl>
#>  1 NA       2022-10-22      wk inc flu … quantile    <NA>                   2380
#>  2 01       2022-10-22      wk inc flu … quantile    NA                      141
#>  3 02       2022-10-22      wk inc flu … quantile    NA                        3
#>  4 04       2022-10-22      wk inc flu … quantile    NA                       22
#>  5 05       2022-10-22      wk inc flu … quantile    NA                       50
#>  6 06       2022-10-22      wk inc flu … quantile    NA                      124
#>  7 08       2022-10-22      wk inc flu … quantile    NA                       15
#>  8 09       2022-10-22      wk inc flu … quantile    NA                        9
#>  9 10       2022-10-22      wk inc flu … quantile    NA                        1
#> 10 11       2022-10-22      wk inc flu … quantile    NA                        8
#> # ℹ 200,330 more rows

# Turn all NA values to "" in output_type_id ----
oracl$output_type_id[is.na(oracl$output_type_id)] <- ""
readr::write_csv(oracl, fs::path(tmp_hub_path, "target-data", "oracle-output.csv"))

# With quoted NA = FALSE, empty strings are not recognised as NAs
arrow::open_dataset(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"),
                    format = "csv",
                    schema = hubData::create_oracle_output_schema(tmp_hub_path),
                    skip = 1) |>
  dplyr::collect()
#> # A tibble: 200,340 × 6
#>    location target_end_date target       output_type output_type_id oracle_value
#>    <chr>    <date>          <chr>        <chr>       <chr>                 <dbl>
#>  1 NA       2022-10-22      wk inc flu … quantile    ""                     2380
#>  2 01       2022-10-22      wk inc flu … quantile    ""                      141
#>  3 02       2022-10-22      wk inc flu … quantile    ""                        3
#>  4 04       2022-10-22      wk inc flu … quantile    ""                       22
#>  5 05       2022-10-22      wk inc flu … quantile    ""                       50
#>  6 06       2022-10-22      wk inc flu … quantile    ""                      124
#>  7 08       2022-10-22      wk inc flu … quantile    ""                       15
#>  8 09       2022-10-22      wk inc flu … quantile    ""                        9
#>  9 10       2022-10-22      wk inc flu … quantile    ""                        1
#> 10 11       2022-10-22      wk inc flu … quantile    ""                        8
#> # ℹ 200,330 more rows

# With quoted NA = TRUE, empty strings recognised as NAs but so is "NA"
arrow::open_dataset(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"),
                    format = "csv",
                    schema = hubData::create_oracle_output_schema(tmp_hub_path),
                    skip = 1, quoted_na = TRUE, na = c("", "NA")) |>
  dplyr::collect()
#> # A tibble: 200,340 × 6
#>    location target_end_date target       output_type output_type_id oracle_value
#>    <chr>    <date>          <chr>        <chr>       <chr>                 <dbl>
#>  1 <NA>     2022-10-22      wk inc flu … quantile    <NA>                   2380
#>  2 01       2022-10-22      wk inc flu … quantile    <NA>                    141
#>  3 02       2022-10-22      wk inc flu … quantile    <NA>                      3
#>  4 04       2022-10-22      wk inc flu … quantile    <NA>                     22
#>  5 05       2022-10-22      wk inc flu … quantile    <NA>                     50
#>  6 06       2022-10-22      wk inc flu … quantile    <NA>                    124
#>  7 08       2022-10-22      wk inc flu … quantile    <NA>                     15
#>  8 09       2022-10-22      wk inc flu … quantile    <NA>                      9
#>  9 10       2022-10-22      wk inc flu … quantile    <NA>                      1
#> 10 11       2022-10-22      wk inc flu … quantile    <NA>                      8
#> # ℹ 200,330 more rows

# With quoted NA = TRUE, and only "" set as the NA value, empty strings recognised 
# as NAs and "NA" value is not.
arrow::open_dataset(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"),
                    format = "csv",
                    schema = hubData::create_oracle_output_schema(tmp_hub_path),
                    skip = 1, quoted_na = TRUE, na = c("")) |>
  dplyr::collect()
#> # A tibble: 200,340 × 6
#>    location target_end_date target       output_type output_type_id oracle_value
#>    <chr>    <date>          <chr>        <chr>       <chr>                 <dbl>
#>  1 NA       2022-10-22      wk inc flu … quantile    <NA>                   2380
#>  2 01       2022-10-22      wk inc flu … quantile    <NA>                    141
#>  3 02       2022-10-22      wk inc flu … quantile    <NA>                      3
#>  4 04       2022-10-22      wk inc flu … quantile    <NA>                     22
#>  5 05       2022-10-22      wk inc flu … quantile    <NA>                     50
#>  6 06       2022-10-22      wk inc flu … quantile    <NA>                    124
#>  7 08       2022-10-22      wk inc flu … quantile    <NA>                     15
#>  8 09       2022-10-22      wk inc flu … quantile    <NA>                      9
#>  9 10       2022-10-22      wk inc flu … quantile    <NA>                      1
#> 10 11       2022-10-22      wk inc flu … quantile    <NA>                      8
#> # ℹ 200,330 more rows
```

<sup>Created on 2025-04-02 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

### Potential solution

The only potential solution I can see here is to:
- continue to set `quoted_na = TRUE` to ensure correct readings of NAs (in connect_hub we do so via `strings_can_be_null = TRUE` which pretty much does the same thing).
- surface the `na` argument to override the default.
- Advise users that need to use an `"NA"` character value in their hub to write their `NA` values as `""` and use the `na = c("")` to specify that.
- This will also need to be available and passed to underlying readers in `hubValidations::read_model_out_file()` as well as `hubData::connect_hub()` and `hubData::connect_model_output()`. It might therefore be prudent to add a section in the schema that allows users to define these settings in their config and all readers to access such info from there.


Note as well that this is the case with other readers so not confined to `arrow` CSV readers:

``` r
tmp_hub_path <- withr::local_tempdir()
example_hub <- "https://github.com/hubverse-org/example-complex-forecast-hub.git"
gert::git_clone(url = example_hub, path = tmp_hub_path)

oracl <-  arrow::read_csv_arrow(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"))
# Add quoted NA value to location
oracl$location[1] <- "NA"
# Turn NA to empty string in output_type_id
oracl$output_type_id[is.na(oracl$output_type_id)] <- ""
readr::write_csv(oracl, fs::path(tmp_hub_path, "target-data", "oracle-output.csv"))


readr::read_csv(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"))
#> Rows: 200340 Columns: 6
#> ── Column specification ────────────────────────────────────────────────────────
#> Delimiter: ","
#> chr  (4): location, target, output_type, output_type_id
#> dbl  (1): oracle_value
#> date (1): target_end_date
#> 
#> ℹ Use `spec()` to retrieve the full column specification for this data.
#> ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
#> # A tibble: 200,340 × 6
#>    location target_end_date target       output_type output_type_id oracle_value
#>    <chr>    <date>          <chr>        <chr>       <chr>                 <dbl>
#>  1 <NA>     2022-10-22      wk inc flu … quantile    <NA>                   2380
#>  2 01       2022-10-22      wk inc flu … quantile    <NA>                    141
#>  3 02       2022-10-22      wk inc flu … quantile    <NA>                      3
#>  4 04       2022-10-22      wk inc flu … quantile    <NA>                     22
#>  5 05       2022-10-22      wk inc flu … quantile    <NA>                     50
#>  6 06       2022-10-22      wk inc flu … quantile    <NA>                    124
#>  7 08       2022-10-22      wk inc flu … quantile    <NA>                     15
#>  8 09       2022-10-22      wk inc flu … quantile    <NA>                      9
#>  9 10       2022-10-22      wk inc flu … quantile    <NA>                      1
#> 10 11       2022-10-22      wk inc flu … quantile    <NA>                      8
#> # ℹ 200,330 more rows
readr::read_csv(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"),
                quoted_na = FALSE)
#> Warning: The `quoted_na` argument of `read_csv()` is deprecated as of readr 2.0.0.
#> This warning is displayed once every 8 hours.
#> Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
#> generated.
#> Rows: 200340 Columns: 6
#> ── Column specification ────────────────────────────────────────────────────────
#> Delimiter: ","
#> chr  (4): location, target, output_type, output_type_id
#> dbl  (1): oracle_value
#> date (1): target_end_date
#> 
#> ℹ Use `spec()` to retrieve the full column specification for this data.
#> ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
#> # A tibble: 200,340 × 6
#>    location target_end_date target       output_type output_type_id oracle_value
#>    <chr>    <date>          <chr>        <chr>       <chr>                 <dbl>
#>  1 <NA>     2022-10-22      wk inc flu … quantile    <NA>                   2380
#>  2 01       2022-10-22      wk inc flu … quantile    <NA>                    141
#>  3 02       2022-10-22      wk inc flu … quantile    <NA>                      3
#>  4 04       2022-10-22      wk inc flu … quantile    <NA>                     22
#>  5 05       2022-10-22      wk inc flu … quantile    <NA>                     50
#>  6 06       2022-10-22      wk inc flu … quantile    <NA>                    124
#>  7 08       2022-10-22      wk inc flu … quantile    <NA>                     15
#>  8 09       2022-10-22      wk inc flu … quantile    <NA>                      9
#>  9 10       2022-10-22      wk inc flu … quantile    <NA>                      1
#> 10 11       2022-10-22      wk inc flu … quantile    <NA>                      8
#> # ℹ 200,330 more rows
readr::read_csv(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"),
                quoted_na = FALSE, na = c(""))
#> Rows: 200340 Columns: 6
#> ── Column specification ────────────────────────────────────────────────────────
#> Delimiter: ","
#> chr  (4): location, target, output_type, output_type_id
#> dbl  (1): oracle_value
#> date (1): target_end_date
#> 
#> ℹ Use `spec()` to retrieve the full column specification for this data.
#> ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
#> # A tibble: 200,340 × 6
#>    location target_end_date target       output_type output_type_id oracle_value
#>    <chr>    <date>          <chr>        <chr>       <chr>                 <dbl>
#>  1 NA       2022-10-22      wk inc flu … quantile    <NA>                   2380
#>  2 01       2022-10-22      wk inc flu … quantile    <NA>                    141
#>  3 02       2022-10-22      wk inc flu … quantile    <NA>                      3
#>  4 04       2022-10-22      wk inc flu … quantile    <NA>                     22
#>  5 05       2022-10-22      wk inc flu … quantile    <NA>                     50
#>  6 06       2022-10-22      wk inc flu … quantile    <NA>                    124
#>  7 08       2022-10-22      wk inc flu … quantile    <NA>                     15
#>  8 09       2022-10-22      wk inc flu … quantile    <NA>                      9
#>  9 10       2022-10-22      wk inc flu … quantile    <NA>                      1
#> 10 11       2022-10-22      wk inc flu … quantile    <NA>                      8
#> # ℹ 200,330 more rows
read.csv(fs::path(tmp_hub_path, "target-data", "oracle-output.csv")) |>
  head()
#>   location target_end_date          target output_type output_type_id
#> 1     <NA>      2022-10-22 wk inc flu hosp    quantile               
#> 2       01      2022-10-22 wk inc flu hosp    quantile               
#> 3       02      2022-10-22 wk inc flu hosp    quantile               
#> 4       04      2022-10-22 wk inc flu hosp    quantile               
#> 5       05      2022-10-22 wk inc flu hosp    quantile               
#> 6       06      2022-10-22 wk inc flu hosp    quantile               
#>   oracle_value
#> 1         2380
#> 2          141
#> 3            3
#> 4           22
#> 5           50
#> 6          124
read.csv(fs::path(tmp_hub_path, "target-data", "oracle-output.csv"),
         na.strings = c("")) |>
  head()
#>   location target_end_date          target output_type output_type_id
#> 1       NA      2022-10-22 wk inc flu hosp    quantile           <NA>
#> 2       01      2022-10-22 wk inc flu hosp    quantile           <NA>
#> 3       02      2022-10-22 wk inc flu hosp    quantile           <NA>
#> 4       04      2022-10-22 wk inc flu hosp    quantile           <NA>
#> 5       05      2022-10-22 wk inc flu hosp    quantile           <NA>
#> 6       06      2022-10-22 wk inc flu hosp    quantile           <NA>
#>   oracle_value
#> 1         2380
#> 2          141
#> 3            3
#> 4           22
#> 5           50
#> 6          124
```

<sup>Created on 2025-04-02 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

zkamvar commented:
Well that's frustrating because [it used to be possible to include both "NA" and NA](https://github.com/tidyverse/readr/issues/295#issuecomment-208398516). I agree then it makes sense to set `quoted_na = TRUE`. FWIW, I attempted to test this using the palmerpenguins data set and I think I got tripped up because I included a factor.

annakrystalli commented:
@zkamvar what do you think about the proposal to surface to `na` argument and document what's required if hubs want to include a character `"NA"` value in their task IDs? Otherwise currently that's not possible.

zkamvar commented:
> @zkamvar what do you think about the proposal to surface to `na` argument and document what's required if hubs want to include a character `"NA"` value in their task IDs? Otherwise currently that's not possible.

I think it makes sense to surface it. It's definitely something for @lshandross to be aware of when they work on hubPredEvalsData.

annakrystalli commented:
Thanks @zkamvar !

OK, I surfaced the `na` argument in the relevant functions and added some function level documentation about it. Sorry I know this is mixing additional things up a bit but it just felt easier to tackle all at once.

I've done a bit of separation in the commits so hopefully it won't be too taxing to review the couple of extra commits?

I've added a simple relevant test about this for `connect_target_oracle_output()` but we could definitely benefit from checking that if using `""` as `NA`s it doesn't affect the schema and that `connect_hub()` also works as expected. Are you ok for me to open an issue to tackle this separately?

zkamvar commented:
> OK, I surfaced the `na` argument in the relevant functions and added some function level documentation about it. Sorry I know this is mixing additional things up a bit but it just felt easier to tackle all at once.

All good! It would be illegal to add a new argument without documenting it anyways. 

> I've added a simple relevant test about this for `connect_target_oracle_output()` but we could definitely benefit from checking that if using `""` as `NA`s it doesn't affect the schema and that `connect_hub()` also works as expected. Are you ok for me to open an issue to tackle this separately?

👍🏽
